### ISSUE TITLE
Renaming ukvm to a more generic name

### ISSUE THREAD

**Issue by dstolfa** [Solo5/solo5#172]
As @mato has submitted a [pull request](https://github.com/Solo5/solo5/pull/171) with abstractions that allow for generalization of the ukvm monitor towards other hypervisor APIs such as Hypervisor.Framework on Mac OS X, libvmmapi(bhyve) on FreeBSD and others, it would make sense to change the name of the monitor to something more generic to prevent confusion on non-KVM hypervisors. A possible name might be `umon` or `ukmon`, however, as @mato has pointed out, renaming ukvm has an impact on downstreams such as MirageOS.

**Comment by amirmc** [comment]
A few folks suggested `uvmm`.  It also happens to sound like 'ovum', meaning 'egg', which I think is nice :)

**Comment by hannesm** [comment]
I talked a bit with @mato in Brussels about renaming -- keep in mind that renaming only ever becomes more expensive if done at a later stage.

The current proposal (please correct me if I missed something, @mato) -- which sidesteps the "find a new name" issue entirely by using `solo5`:
- in mirage/mirage: rename `ukvm` to `solo5` as target (leave `muen` and `virtio` as-is)
- in opam-repository: rename `solo5-kernel-ukvm` to `solo5-kernel` (as the package name), leave others as they are
- in solo5 codebase: rename `ukvm-bin` to `solo5`

/cc @mato @djwillia @ricarkol  (please comment here if there's an argument against the above)

**Comment by djwillia** [comment]
That sounds about right to me.  The way I think about it is that what we were calling solo5 are just the bindings for what we were calling ukvm.  As the solo5 api (solo5.h) and the formerly-known-as-ukvm hypercall interface converge (e.g., PR 224 moving malloc is a step toward that) , it  should be more obvious that they are the same thing, hence only needing a single name.  And the word "solo5" does not have the KVM baggage and association that the word "ukvm" does. 

The other slightly confusing part is that there are "other platforms" that can be accessed through the solo5 api (muen and virtio).  For now we are calling them "compatibility".

**Comment by ricarkol** [comment]
Sounds good @hannesm.

Regarding the WIP monitors: seccomp and SGX. What about naming those `solo5-seccomp` and `solo5-sgx`? this would mean that the default `solo5` is the one that uses KVM or VMM.

**Comment by mato** [comment]
The missing piece in this approach is that it leaves the "Solo5 { monitor | bindings } to hardware virtualization" without a name.

Thanking about it, what we want is concise and consistent names for:

1. The Solo5 API bindings to _X_ (what is known in OPAM as `solo5-kernel-X`, and what we have in the repository as `kernel/X`).
2. The Solo5 loader/monitor for _X_ (what we have in the repository as `ukvm/` and compiles to `ukvm-bin`).
3. MirageOS targets for _X_ (what the user specifies as `mirage configure -t X`).

In all cases I think there will be more combinations in the future, and there will be other virtio- or muen- like _X_, so I don't want to call them "compatibility" layers.

Here's a scheme which could work for all of the above for the next release:

1. The Solo5 API bindings to _X_ live in `bindings/x`, and are packaged in OPAM as `solo5-kernel-x`.
2. The Solo5 loader/monitor for _X_ (if there is one) lives in `monitors/x`, compiles to `solo5-x` _and_ at least until we change the current build-the-monitor-from-source-at-application-build-time scheme are included in OPAM in the `solo5-kernel-x` package.
3. The MirageOS target for _X_ is just `mirage configure -t x`.

Where _X_ is one of:

- `vt` (formerly known as `ukvm`)
- `virtio`
- `muen`
- (future) `sgx`, `seccomp`, ...

If I wanted to get _really_ pedantic about it I'd also rename the OPAM `solo5-kernel-x` to `solo5-bindings-x` post 0.3.0 but I'm not sure it's worth the hassle. Also, I have this nagging feeling that we could actually combine those OPAM packages into a single package, so I'd rather wait some more until that's clear.

What do you think?

**Comment by hannesm** [comment]
@mato I like your proposal! Thanks

**Comment by djwillia** [comment]
I think there is some similarity shared between `vt` and future `sgx`, `seccomp`, etc. that is not shared with `virtio` and `muen`.

For example, I think `solo5-x` makes sense for the first group, but `solo5-virtio` doesn't really make sense as there is no (and will never be) a "solo5 monitor" for virtio (e.g., QEMU is one monitor for virtio, but has nothing to do with solo5).  For muen, I could imagine a `solo5-muen` that perhaps runs as a different muen subject connected to the unikernel, but the existing "monitor" for the muen backend is the muen separation kernel itself, which again does not have anything to do with solo5.  On the other hand, for `solo5-vt` the "monitor" does have something to do with solo5 in that it more or less directly packages the solo5 API as hypercalls and I think in both the future `sgx` and `seccomp` it will do a direct packaging as well.

In other words, in the current `virtio` and `muen` cases, `kernel/` is not acting like bindings, but more like a translation or compatibility layer to a different "monitor" that is not related to Solo5.

Does that make sense?

**Comment by mato** [comment]
> In other words, in the current virtio and muen cases, kernel/ is not acting like bindings, but more like a translation or compatibility layer to a different "monitor" that is not related to Solo5.

Well, the world is not black and white. We need to invent and/or appropriate terminology which is good enough for the components we have, primarily so that _we_ know what we're talking about and can clearly explain it to newcomers in documentation, talks, papers or the code itself.

In my opinion the term _bindings_ clearly refers to the layer that binds/translates the Solo5 API to _X_. You can think of the different cases as "fatter" (`virtio`) or "thinner" (`vt`) bindings if that helps. It's certainly clearer than _kernel_ which brings other connotations with it, much like "virtual machine" that you wanted to get rid of for the monitor.

Regarding the monitor, I still have an issue with that term, since for some _X_ there will be a component in that position but it will _not_ do any monitoring. During earlier discussions face to face  we variously danced around a _loader or monitor_ combination, which seems like a cop out.

I went hunting for more terms to appropriate, once nice possibility is  _tender_:

> _tender_
> noun
> 1. with modifier A vehicle used by a fire service for carrying specified supplies or equipment or fulfilling a specified role. _‘three fire engines, including an emergency tender, attended the scene’_
>    1.1 A vehicle used in mobile operations by a public service or the armed forces. _‘she saw the police tender dash up and half a dozen men tumble out’_
> 2. A dinghy or other boat used to ferry people and supplies to and from a ship.
> 3. A trailing vehicle closely coupled to a steam locomotive to carry fuel and water.
> 4. usually in combination or with modifier A person who looks after someone else or a machine or place. _‘Alexei signalled to one of the engine tenders’_
> 
> Origin
>
> Late Middle English (in the sense ‘attendant, nurse’): from tend or shortening of attender (see attend).

This is a much looser term than monitor but still has the association of _tending to the guest_ in some unspecified way. It also has the advantage of not being previously used in the systems/software engineering domain.

What do you think?

**Comment by mato** [comment]
@ricarkol @djwillia @hannesm @dstolfa Any opinions on "monitor" vs. "tender" as I describe it above? I'm more or less happy to use either, but would like to settle the terminology.

**Comment by hannesm** [comment]
I like `tender` since it is fresh (so Solo5 can define what it means), but would be happy with `monitor` as well.

**Comment by ricarkol** [comment]
I vote for `tender`. Calling them monitors might be more confusing (specially the seccomp one).

**Comment by djwillia** [comment]
OK I'm in agreement that the terminology you suggested is good for our purposes at this time.  Also, I'm happy with `tender` and it will be interesting to see what some of the cool new ways of "tending to the guest" come up :)

**Comment by mato** [comment]
So, `tender` it is then. I will do this renaming last, after getting through the various PRs in progress.

**Comment by hannesm** [comment]
this can be closed now :)

**Comment by mato** [comment]
All done here, closing!
